vespalog/src/vespa/log/bufferedlogger.cpp

```diff
@@ -246,143 +255,113 @@ BufferedLogger::doLog(Logger& l, Logger::LogLevel level,
+    Entry &entry = _cache.getOrAdd(newEntry);
+    if (entry._count++ == 0) {
+        // If entry didn't already exist, log it
+        TimeStampWrapper wrapper(entry.payload.timestamp);
+        entry.log(wrapper, message);
     }
-    trimCache(entry._timestamp);
+    trimCache(entry.payload.timestamp);
+
 }
 
-void
-BackingBuffer::flush()
-{
+void BackingBuffer::flush() {
     std::lock_guard<std::mutex> guard(_mutex);
-    for (const auto & entry : _cacheBack) {
-        log(entry);
+    for (const auto & entry : _cache) {
+        logIfRepeated(entry);
     }
-    _cacheBack.clear();
-    for (const auto & entry : _cacheFront) {
-        log(entry);
-    }
-    _cacheFront.clear();
+    _cache.clear();
 }
 
-void
-BufferedLogger::flush() {
+void BufferedLogger::flush() {
     _backing->flush();
 }
 
-void
-BackingBuffer::trimCache(system_time currentTime)
-{
+void BackingBuffer::trimCache(system_time currentTime) {
+    std::vector<uint64_t> removeList;
+    for (const auto & entry : _cache) {
         // Remove entries that have been in here too long.
-    while (!_cacheBack.empty() &&
-           _cacheBack.front()._timestamp + _maxEntryAge < currentTime)
-    {
-        log(_cacheBack.front());
-        _cacheBack.pop_front();
+        if (entry.payload.timestamp + _maxEntryAge < currentTime) {
+            logIfRepeated(entry);
+            // can't remove() while iterating
+            removeList.push_back(entry.sequenceId);
+        } else {
+            // no more old entries
+            break;
+        }
     }
-    while (!_cacheFront.empty() &&
-           _cacheFront.front()._timestamp + _maxEntryAge < currentTime)
-    {
-        log(_cacheFront.front());
-        _cacheFront.pop_front();
+    for (uint64_t sequenceId : removeList) {
+        _cache.remove(sequenceId);
     }
-        // If cache front is larger than half max size, move to back.
-    for (uint32_t i = _cacheFront.size(); i > _maxCacheSize / 2; --i) {
-        Entry e(_cacheFront.front());
-        _cacheFront.pop_front();
-        _cacheBack.push_back(e);
-    }
-        // Remove entries from back based on count modified age.
-    for (uint32_t i = _cacheFront.size() + _cacheBack.size(); i > _maxCacheSize; --i) {
-        log(*_cacheBack.get<2>().begin());
-        _cacheBack.get<2>().erase(_cacheBack.get<2>().begin());
+    // could really be if, since we trim each time we add to cache
+    while (_cache.size() > _maxCacheSize) {
+        // the newest entries are immune to removal:
+        size_t immune = (_maxCacheSize / 2);
+        size_t numToCheck = _cache.size() - immune;
+        system_time oldest;
+        const Entry *toRemove = nullptr;
+        for (const auto & entry : _cache) {
+            system_time adjusted = entry.getAgeFactor();
+            if (toRemove == nullptr || adjusted < oldest) {
+                oldest = adjusted;
+                toRemove = &entry;
+            }
+            if (--numToCheck == 0) break;
+        }
```

Comment:
Although not likely an issue in practice since we're generally dealing with small log caches, we may consider ensuring sub-linear complexity for finding the eviction candidate when the cache is at capacity.